calc/vxlan_resolver.go

```diff
@@ -0,0 +1,453 @@
+package calc
+
+import (
+	"crypto/sha1"
+	gonet "net"
+
+	"github.com/projectcalico/felix/dispatcher"
+	"github.com/projectcalico/felix/ip"
+	"github.com/projectcalico/felix/proto"
+	"github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/encap"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"github.com/projectcalico/libcalico-go/lib/net"
+	"github.com/projectcalico/libcalico-go/lib/set"
+	"github.com/sirupsen/logrus"
+)
+
+func NewVXLANResolver(hostname string, callbacks PipelineCallbacks) *VXLANResolver {
+	return &VXLANResolver{
+		hostname:                  hostname,
+		callbacks:                 callbacks,
+		nodeNameToVXLANTunnelAddr: map[string]string{},
+		nodeNameToIPAddr:          map[string]string{},
+		nodeNameToSentRoutes:      map[string]set.Set{},
+		blockToRoutes:             map[string]map[string]*vxlanRoute{},
+		vxlanPools:                map[string]model.IPPool{},
+		pendingRouteUpdates:       set.New(),
+		sentVTEPs:                 map[string]bool{},
+	}
+}
+
+// VXLANResolver is responsible for resolving node IPs, node config, IPAM blocks,
+// and IP pools to determine the correct set of programming to send to the dataplane.
+// Specifically, it registers to get updates on the following:
+//
+//   - model.HostIPKey
+//   - model.HostConfigKey
+//   - model.BlockKey
+//   - model.IPPoolKey
+//
+// VXLAN routes are contributed by blocks within IP pools that have VXLAN enabled, and
+// must target a VXLAN tunnel endpoint (VTEP) which comprises a node IP address, VXLAN
+// tunnel address, and a deterministically calculated MAC address. The VXLAN resolver
+// ensurs that routes are only sent to the data plane when a fully specified VTEP
+// exists.
+//
+// For each VTEP, this component will send a *proto.VXLANTunnelEndpointUpdate followed by
+// a *proto.RouteUpdate for each route which targets that VTEP. As routes are added and
+// removed, subsequent *proto.RouteUpdate and *proto.RouteRemove messages will be sent.
+//
+// If a VTEP is no longer fully specified (e.g., due to a vxlan tunnel address removal),
+// a *proto.RouteRemove message will be sent for each route targeting that VTEP, followed
+// by a *proto.VXLANTunnelEndpointRemove message for the VTEP itself.
+//
+// If a VTEP changes (e.g., due to a vxlan tunnel address changing), this component will treat
+// it as a delete followed by an add.
+type VXLANResolver struct {
+	hostname  string
+	callbacks PipelineCallbacks
+
+	// pendingRouteUpdates stores the set of route updates which
+	// we may want to program in the future, but cannot yet because
+	// they are not yet valid (e.g., due to a missing VTEP).
+	pendingRouteUpdates set.Set
```

Comment:
Wondering if we can do away with the pending and sent sets here by calculating whether the route would already have been sent.

Reply:
Sounds good to me so long as we're good with iterating all the routes more frequently. We already have the potential to iterate lots of routes with the current implementation anyway, so I'll see what it will take to adjust to something like this.